ads/adform.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {writeScript, validateSrcPrefix, checkData} from '../src/3p';
+
+// Valid adform ad source hosts
+const hosts = {
+  track: 'https://track.adform.net/',
+  adx: 'https://adx.adform.net/'
+};
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function adform(global, data) {
+
+  checkData(data, ['src', 'bn', 'mid']);
+  global.Adform = {ampData: data};
+  const src = data.src;
+  const bn = data.bn;
+  const mid = data.mid;
+  let url;
+
+  // Custom ad url using "data-src" attribute
+  if (src) {
+
+    let lastError;
+
+    // NOTE: Since validateSrcPrefix() throws and we need to validate src to be
+    // from one of the two whitelisted hosts - we will look for two exceptions
+    // and re-throw validation error with invalid "data-src" attribute.
+    for (const type in hosts) {
+
+      try {
+        validateSrcPrefix(hosts[type], src);
```

Comment:
Would you mind instead changing `validateSrcPrefix` to take an array of prefixes? And add some unit tests there?